File: 程式設計：資料結構答案檔/C#/503.cs
using System;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        string[] firstLine = Console.ReadLine().Split(' ');
        int n1 = int.Parse(firstLine[0]);
        int n2 = int.Parse(firstLine[1]);

        List<int> A = new List<int>(new int[n1 + 1]);
        List<int> B = new List<int>(new int[n2 + 1]);

        string[] AInputs = Console.ReadLine().Split(' ');
        for (int i = 1; i <= n1; i++)
        {
            A[i] = int.Parse(AInputs[i - 1]);
        }

        string[] BInputs = Console.ReadLine().Split(' ');
        for (int j = 1; j <= n2; j++)
        {
            B[j] = int.Parse(BInputs[j - 1]);
        }

        List<int> DP = new List<int>(new int[n2 + 1]);

        for (int i = 1; i <= n1; i++)
        {
            List<int> temp = new List<int>(new int[n2 + 1]);
            for (int j = 1; j <= n2; j++)
            {
                if (A[i] != B[j])
                {
                    temp[j] = Math.Max(temp[j - 1], DP[j]);
                }
                else
                {
                    temp[j] = DP[j - 1] + 1;
                }
            }
            DP = temp;
        }

        Console.WriteLine(DP[n2]);
    }
}

File: 程式設計：資料結構答案檔/Python/503.py
n1, n2 = map(int, input().split())
A = [0]
A.extend(map(int, input().split()))
B = [0]
B.extend(map(int, input().split()))

DP = [0 for i in range(n2+1)]
for i in range(1, n1+1):
    temp = [0 for i in range(n2+1)]
    for j in range(1, n2+1):
        if A[i] != B[j]: temp[j] = max([temp[j-1], DP[j]])
        else: temp[j] = DP[j-1]+1
    DP = temp;

print(DP[n2])
